File: modules/core/app/model/blanks/BlanksExercise.java
package model.blanks;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import model.exercise.Exercise;
import model.result.SuccessType;
import play.twirl.api.Html;

public class BlanksExercise extends Exercise {

  private final List<BlankObject> objects;

  public BlanksExercise(int id) {
    super(id);
    objects = Arrays.asList(new BlankObject(1,
        "&lt;?xml version=\"1.0\" encoding=\"UTF-8\"?&gt;\n&lt;!DOCTYPE root SYSTEM \"doctype.dtd\"&gt;\n&lt;", 4,
        "&gt;", "root"), new BlankObject(2, "&lt;", 5, "&gt;", "/root"));
  }

  public List<SuccessType> correct(List<String> inputs) {
    if(inputs.size() != objects.size())
      throw new IllegalArgumentException("Es waren zu viele oder zu wenige Input-Elemente!");

    final List<SuccessType> results = new ArrayList<>(inputs.size());

    for(int i = 0; i < inputs.size(); i++)
      results.add(inputs.get(i).equals(objects.get(i).getSolution()) ? SuccessType.COMPLETE : SuccessType.NONE);

    return results;
  }

  public List<BlankObject> getObjects() {
    return objects;
  }

  public Html render() {
    return new Html(objects.stream().map(BlankObject::render).collect(Collectors.joining("\n")));
  }

}

File: modules/core/app/model/blanks/BlanksExercise.scala
package model.blanks


import model.exercise.Exercise
import model.result.SuccessType
import play.twirl.api.Html

class BlanksExercise extends Exercise {

  val objects: List[BlankObject] = List(
    new BlankObject(1,
      """&lt?xml version="1.0" encoding="UTF-8"?&gt"
        |&lt!DOCTYPE root SYSTEM "doctype.dtd"&gt
        |&lt""".stripMargin, 4, "&gt", "root"),
    new BlankObject(2, "&lt", 5, "&gt", "/root")
  )

  def correct(inputs: List[String]): List[SuccessType] =
    for {(inp, obj) <- inputs zip objects}
      yield if (inp == obj.getSolution) SuccessType.COMPLETE else SuccessType.NONE


  def render = new Html(objects.map(_.render).mkString("\n"))

}
